src/cs/Ssh.Tcp/Services/PortForwardingService.cs

```diff
@@ -893,53 +955,57 @@ await remoteConnector.OnChannelOpeningAsync(request, cancellation)
 		return channel;
 	}
 
-	internal void AddChannelForwarder(ChannelForwarder channelForwarder)
+	internal void AddStreamForwarder(StreamForwarder streamForwarder)
 	{
-		if (this.disposed)
+		lock (DisposablesLock)
 		{
-			channelForwarder.Dispose();
-		}
-		else
-		{
-			lock (this.channelForwarders)
+			if (this.disposed)
+			{
+				streamForwarder.Dispose();
+			}
+			else
 			{
-				this.channelForwarders.Add(channelForwarder);
+				this.streamForwarders.Add(streamForwarder);
 			}
 		}
-	}
 
-	internal void RemoveChannelForwarder(ChannelForwarder channelForwarder)
-	{
-		lock (this.channelForwarders)
+		streamForwarder.Closed += (_, _) =>
 		{
-			this.channelForwarders.Remove(channelForwarder);
-		}
+			lock (DisposablesLock)
+			{
+				this.streamForwarders.Remove(streamForwarder);
+			}
+		};
 	}
 
 	protected override void Dispose(bool disposing)
 	{
 		if (disposing && !this.disposed)
 		{
-			this.disposed = true;
-
 			var disposables = new List<IDisposable>();
 
-			lock (this.channelForwarders)
+			lock (DisposablesLock)
 			{
-				disposables.AddRange(this.channelForwarders);
-				this.channelForwarders.Clear();
-			}
+				if (this.disposed) return;
+				this.disposed = true;
 
-			lock (this.localForwarders)
-			{
-				disposables.AddRange(this.localForwarders.Values);
-				this.localForwarders.Clear();
-			}
+				lock (this.streamForwarders)
```

Comment:
It's already locked on line 987: `lock (DisposablesLock)`

Reply:
Removed.